Add Tesla Fleet Energy History

JEMcats opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project! #You're Welcome!#
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This pull request adds 7 features:
- Solar Production For The Day ('kWh')
- Grid Import For The Day ('kWh')
- Grid Export For The Day ('kWh')
- Battery Charged For The Day ('kWh')
- Battery Discharged For The Day ('kWh')

and resolves a [feature request](https://community.home-assistant.io/t/tesla-fleet-api-and-energy-tab/)!


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests


## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works. <!-- I'm not sure what this means or how to do this but, I have tested everything in a local instance of homeassistant and it appears to work correctly -->

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository] The updated documentation is here:
[tesla_fleet.markdown.zip](https://github.com/user-attachments/files/17006606/tesla_fleet.markdown.zip)

   
If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.  <!-- There are no new dependencies -->
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description. <!-- There are no new dependencies -->

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

JEMcats committed:
Add "Grid Status" and "Storm Watch active" values.

JEMcats committed:
Adding Sensors For Energy History

JEMcats committed:
Update Naming

JEMcats committed:
updated requirements

JEMcats commented:
Hello, currently the [python-tesla-fleet-api](https://github.com/Teslemetry/python-tesla-fleet-api) package used in the Tesla Fleet API integration has a bug that would stop this update from working, i have submitted a [pull request](https://github.com/Teslemetry/python-tesla-fleet-api/pull/3) to [python-tesla-fleet-api](https://github.com/Teslemetry/python-tesla-fleet-api) that resolves this issue but until that pull request is merged by [python-tesla-fleet-api](https://github.com/Teslemetry/python-tesla-fleet-api), **DO NOT** merge this update to the Fleet API integration with homeassistant core. I will add a new comment here if [python-tesla-fleet-api](https://github.com/Teslemetry/python-tesla-fleet-api) merges my pull request. Thank you in advance.

JEMcats committed:
Merge branch 'home-assistant:dev' into tesla-solar

Bre77 commented on homeassistant/const.py:
```diff
@@ -654,6 +654,8 @@ class UnitOfPower(StrEnum):
     """Power units."""
 
     WATT = "W"
+    WATT_HOUR = "Wh"
```
This is not a unit of Power, it's a Unit of Energy, which should already exists. I don't think you're allowed to edit this file.

JEMcats replied:
I just pushed a commit that fixes this.

Bre77 commented on requirements_all.txt:
```diff
@@ -2776,11 +2776,13 @@ temperusb==1.6.1
 tesla-fleet-api==0.7.3
 
+# homeassistant.components.tesla_fleet
+tesla-fleet-api==0.7.4
```
You can't bump just one integration, you have to bump Tessie and Teslemetry too.

JEMcats replied:
I just pushed a commit that fixes this.

Bre77 reviewed (changes requested):
We will also need tests, and I've called out a few issues. Happy to work with you to fix any issues with the library for that endpoint.

JEMcats committed:
Do suggested changes.

JEMcats committed:
Change sensors from Measurement to Total.

JEMcats commented:
If I misunderstood something or you need anything else then feel free to reach out.

Bre77 commented on homeassistant/components/tesla_fleet/coordinator.py:
```diff
@@ -182,6 +182,80 @@ async def _async_update_data(self) -> dict[str, Any]:
+            hass,
+            LOGGER,
+            name="Tesla Fleet Energy Site Past",
+            update_interval=timedelta(seconds=10),
```
The data updates every 5 minutes, polling every 10 seconds seems overkill. I would make this 1 minute at a minimum, but given Tesla's harsh rate limits was thinking 5 minutes may actually be appropriate.

JEMcats replied:
I would be happy to change that this afternoon, I would just like to point out that to my knowledge this endpoint is not rate limited.

![image](https://github.com/user-attachments/assets/0a8e698d-6140-4f75-856d-604bb6af7773)

JEMcats replied:
Changed To 5 Minutes (300 Seconds)

Bre77 commented on homeassistant/components/tesla_fleet/coordinator.py:
```diff
@@ -182,6 +182,80 @@ async def _async_update_data(self) -> dict[str, Any]:
+            "battery_out": 0,
+        }
+        for time_entity in data["time_series"]:
+            formatted_data["grid_in"] = (
+                formatted_data["grid_in"] + time_entity["grid_energy_imported"]
+            )
```
```suggestion
            formatted_data["grid_in"] += time_entity["grid_energy_imported"]
```

Bre77 commented on homeassistant/components/tesla_fleet/coordinator.py:
```diff
@@ -182,6 +182,80 @@ async def _async_update_data(self) -> dict[str, Any]:
+            formatted_data["grid_in"] = (
+                formatted_data["grid_in"] + time_entity["grid_energy_imported"]
+            )
+            formatted_data["grid_out"] = (
+                formatted_data["grid_out"]
+                + time_entity["grid_energy_exported_from_solar"]
+            )
+            +time_entity["grid_energy_exported_from_generator"]
+            +time_entity["grid_energy_exported_from_battery"]
```
```suggestion
            formatted_data["grid_out"] += (
            time_entity["grid_energy_exported_from_solar"]
            + time_entity["grid_energy_exported_from_generator"]
            + time_entity["grid_energy_exported_from_battery"]
            )
```

Bre77 commented on homeassistant/components/tesla_fleet/manifest.json:
```diff
@@ -8,5 +8,5 @@
   "loggers": ["tesla-fleet-api"],
   "quality_scale": "gold",
-  "requirements": ["tesla-fleet-api==0.7.3"]
+  "requirements": ["tesla-fleet-api==0.7.4"]
```
```suggestion
  "requirements": ["tesla-fleet-api==0.7.8"]
```
Might have to do this in a seperate PR so that it also bumps Tessie and Teslemetry

JEMcats replied:
I was able to update Tesla Fleet to 0.7.8 but not Tessie and Teslemetry.

Bre77 commented on homeassistant/components/tesla_fleet/sensor.py:
```diff
@@ -439,6 +491,12 @@ async def async_setup_entry(
+                TeslaFleetEnergyPastSensorEntity(energysite, description)
+                for energysite in entry.runtime_data.energysites
+                for description in ENERGY_PAST_DESCRIPTIONS
+                if description.key in energysite.past_coordinator.data
```
```suggestion
                if energysite.info_coordinator.data.get("components_battery")
                or energysite.info_coordinator.data.get("components_solar")
```

Bre77 replied:
Users with only a powerwall cant use this API so no point even trying to add it.

Bre77 commented on homeassistant/components/tesla_fleet/sensor.py:
```diff
@@ -540,6 +598,33 @@ def _async_update_attrs(self) -> None:
+
+    def _async_update_attrs(self) -> None:
+        """Update the attributes of the sensor."""
+        self._attr_available = not self.is_none
```
```suggestion
```

I dont think its possible for the values to be None

Bre77 commented on homeassistant/components/tesla_fleet/sensor.py:
```diff
@@ -378,6 +379,14 @@ class TeslaFleetTimeEntityDescription(SensorEntityDescription):
         device_class=SensorDeviceClass.POWER,
         entity_registry_enabled_default=False,
     ),
+    SensorEntityDescription(
+        key="island_status",
+        entity_registry_enabled_default=True,
+    ),
```
This was previously rejected because it didnt have the possible values or translations, which would make it a breaking change. We might be able to add it now because I think I do know all possible values, but recommend it be done in a seperate PR.

See for context: https://github.com/Teslemetry/hass-teslemetry/compare/v2.0.11...v2.0.12

JEMcats replied:
I believe you are correct, to the best of my knowledge the states are:
```
{
          "island_status_unknown": "Unknown",
          "on_grid": "On grid",
          "off_grid": "Off grid",
          "off_grid_intentional": "Off grid intentionally"
}
```

JEMcats replied:
Removed and new pull request made. [#126438](https://github.com/home-assistant/core/pull/126438)

Bre77 commented on homeassistant/components/tesla_fleet/coordinator.py:
```diff
@@ -182,6 +182,80 @@ async def _async_update_data(self) -> dict[str, Any]:
+        self.update_interval = ENERGY_INTERVAL
+        n = datetime.now(UTC).isoformat()
+        try:
+            data = (await self.api.energy_history("energy", n, n, "day", "UTC"))[
```
The parameters for this have changed. I recommend just pulling 1 day in the installations timezone like this.

```suggestion
            data = (await self.api.energy_history("day"))[
```

Bre77 reviewed (changes requested):
I dont love calling it `past`, because the API endpoint is `history`, but thats minor. Lots of little things to consider.

JEMcats committed:
Update coordinator.py

Co-authored-by: Brett Adams <Bre77@users.noreply.github.com>

JEMcats committed:
Update coordinator.py

Co-authored-by: Brett Adams <Bre77@users.noreply.github.com>

JEMcats committed:
Merge branch 'home-assistant:dev' into tesla-solar

JEMcats committed:
Update requirements, change formatting, change history update interval to 5 minutes (300 seconds), and change past to history.

JEMcats committed:
Merge branch 'tesla-solar' of https://github.com/JEMcats/ha-core into tesla-solar

JEMcats committed:
Merge branch 'home-assistant:dev' into tesla-solar

JEMcats committed:
Add more details to Energy History.

JEMcats committed:
Merge branch 'home-assistant:dev' into tesla-solar

JEMcats commented:
Good Afternoon, I have updated the code to follow your recommendations and also added more functionality from the energy history endpoint, I have also updated the markdown file for home-assistant.io which you can find here: 
[tesla_fleet.markdown.zip](https://github.com/user-attachments/files/17035130/tesla_fleet.markdown.zip)

JEMcats committed:
Merge branch 'home-assistant:dev' into tesla-solar

Bre77 commented on tests/components/tesla_fleet/fixtures/history_status.json:
```diff
@@ -0,0 +1,45 @@
+    "period": "day",
+    "time_series": [
+      {
+        "timestamp": "2023-06-01T01:00:00-07:00",
+        "solar_energy_exported": 70940,
+        "generator_energy_exported": 0,
+        "grid_energy_imported": 521,
+        "grid_services_energy_imported": 17.53125,
+        "grid_services_energy_exported": 3.80859375,
+        "grid_energy_exported_from_solar": 43660,
+        "grid_energy_exported_from_generator": 0,
+        "grid_energy_exported_from_battery": 19,
+        "battery_energy_exported": 10030,
+        "battery_energy_imported_from_grid": 80,
+        "battery_energy_imported_from_solar": 16800,
+        "battery_energy_imported_from_generator": 0,
+        "consumer_energy_imported_from_grid": 441,
+        "consumer_energy_imported_from_solar": 10480,
+        "consumer_energy_imported_from_battery": 10011,
+        "consumer_energy_imported_from_generator": 0
```
This fixture appears to be missing the `total_*` values which are only returned when their values are >0, is that intentional?

JEMcats replied:
That's interesting, I think that Tesla has updated their Fleet API from when they published their documentation because that test file is just copied from the documentation with values changed, but it shouldn't matter because those values are not used in the integration.

Bre77 replied:
I guess my point is if we use those values it greatly simplifys things because you dont need to add all the other values together and can use the totals directly.

JEMcats committed:
Merge branch 'home-assistant:dev' into tesla-solar

JEMcats committed:
Merge branch 'home-assistant:dev' into tesla-solar

JEMcats committed:
Merge branch 'home-assistant:dev' into tesla-solar

JEMcats commented:
I’ve been trying to run Home Assistant Core from the [JEMcats:tesla-solar](https://github.com/JEMcats/ha-core/tree/tesla-solar) branch, but I’m getting an error. 
```
2024-09-21 12:37:11.256 ERROR (MainThread) [root] Uncaught exception
Traceback (most recent call last):
  File "/home/vscode/.local/ha-venv/bin/hass", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/workspaces/ha-core/homeassistant/__main__.py", line 209, in main
    exit_code = runner.run(runtime_conf)
                ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/workspaces/ha-core/homeassistant/runner.py", line 189, in run
    return loop.run_until_complete(setup_and_run_hass(runtime_config))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/asyncio/base_events.py", line 687, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "/workspaces/ha-core/homeassistant/runner.py", line 157, in setup_and_run_hass
    hass = await bootstrap.async_setup_hass(runtime_config)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/workspaces/ha-core/homeassistant/bootstrap.py", line 325, in async_setup_hass
    await async_from_config_dict(config_dict, hass) is not None
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/workspaces/ha-core/homeassistant/bootstrap.py", line 457, in async_from_config_dict
    await async_load_base_functionality(hass)
  File "/workspaces/ha-core/homeassistant/bootstrap.py", line 426, in async_load_base_functionality
    await asyncio.gather(
  File "/workspaces/ha-core/homeassistant/helpers/system_info.py", line 53, in async_get_system_info
    hassio = await async_import_module(hass, "homeassistant.components.hassio")
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/workspaces/ha-core/homeassistant/helpers/importlib.py", line 60, in async_import_module
    import_future.result()
  File "/workspaces/ha-core/homeassistant/helpers/importlib.py", line 50, in async_import_module
    module = await hass.async_add_import_executor_job(_get_module, cache, name)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/workspaces/ha-core/homeassistant/helpers/importlib.py", line 26, in _get_module
    cache[name] = importlib.import_module(name)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/workspaces/ha-core/homeassistant/util/loop.py", line 200, in protected_loop_func
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/importlib/__init__.py", line 90, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1387, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1331, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 935, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 995, in exec_module
  File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
  File "/workspaces/ha-core/homeassistant/components/hassio/__init__.py", line 50, in <module>
    from . import (  # noqa: F401
  File "/workspaces/ha-core/homeassistant/components/hassio/binary_sensor.py", line 17, in <module>
    from .entity import HassioAddonEntity
  File "/workspaces/ha-core/homeassistant/components/hassio/entity.py", line 24, in <module>
    from .coordinator import HassioDataUpdateCoordinator
  File "/workspaces/ha-core/homeassistant/components/hassio/coordinator.py", line 10, in <module>
    from aiohasupervisor import SupervisorError
ModuleNotFoundError: No module named 'aiohasupervisor'
```
I’ve also tried running it from the [JEMcats:dev](https://github.com/JEMcats/ha-core/tree/dev) branch, which is up to date with the [home-assistant:dev](https://github.com/home-assistant/core/tree/dev) branch, but the issue continues. I’m not sure what’s causing the issue, but I’d appreciate any suggestions you might have. I do not believe that this issue is part of the Tesla_fleet integration.

JEMcats committed:
Merge branch 'home-assistant:dev' into tesla-solar

JEMcats commented:
I removed the dev container and recreated it, which seems to have fixed that issue.

Bre77 commented:
@JEMcats I am going to raise some PRs against your fork to make this PR look more like https://github.com/home-assistant/core/pull/126166.

I am going to recommend you remove the grid and storm watch sensors because they should go into a seperate PR.

JEMcats committed:
Merge branch 'home-assistant:dev' into tesla-solar

JEMcats committed:
Remove Island Status and Storm Watch Active.

JEMcats commented:
Hi @Bre77, I have remove the Grid Status and Storm Watch Active sensors from this PR and made a new one here: [#126438](https://github.com/home-assistant/core/pull/126438)

JEMcats committed:
Use total_home_usage and total_grid_energy_exported values.

JEMcats committed:
Fix startup bug.

JEMcats committed:
Merge branch 'dev' into tesla-solar

JEMcats commented:
I have updated the code to use the total_home_usage and total_grid_energy_exported values. [3739299](https://github.com/home-assistant/core/pull/126003/commits/3739299deced272ab767c4849cb792d28844e729) and [875838f](https://github.com/home-assistant/core/pull/126003/commits/875838f10966b2b8f9a0dd56713bac99337b0228)

JEMcats committed:
Merge branch 'home-assistant:dev' into tesla-solar
